From 1093a730b1fd161898dba684e99f0ecebc22d932 Mon Sep 17 00:00:00 2001
From: Tomasz Duszynski <tomasz.duszynski@cavium.com>
Date: Wed, 7 Nov 2018 14:02:20 +0100
Subject: [PATCH 0754/1051] soc: octeontx2: do not alloc RXVLAN for LBK VFs

LKB VFs do not have MCAM UCAST entry hence do not alloc
RXVLAN for them.

Signed-off-by: Tomasz Duszynski <tduszynski@marvell.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 drivers/soc/marvell/octeontx2/rvu_nix.c | 15 ++++++---------
 drivers/soc/marvell/octeontx2/rvu_npc.c |  2 +-
 2 files changed, 7 insertions(+), 10 deletions(-)

diff --git a/drivers/soc/marvell/octeontx2/rvu_nix.c b/drivers/soc/marvell/octeontx2/rvu_nix.c
index bfff6291627d..30a2acdd14a3 100644
--- a/drivers/soc/marvell/octeontx2/rvu_nix.c
+++ b/drivers/soc/marvell/octeontx2/rvu_nix.c
@@ -2514,12 +2514,17 @@ int rvu_mbox_handler_NIX_RXVLAN_ALLOC(struct rvu *rvu, struct msg_req *req,
 {
 	struct npc_mcam_alloc_entry_req alloc_req = { };
 	struct npc_mcam_alloc_entry_rsp alloc_rsp = { };
-	struct npc_mcam_ena_dis_entry_req ena_req = { };
 	struct npc_mcam_free_entry_req free_req = { };
 	u16 pcifunc = req->hdr.pcifunc;
 	int blkaddr, nixlf, err;
 	struct rvu_pfvf *pfvf;
 
+	/* LBK VFs do not have separate MCAM UCAST entry hence
+	 * skip allocating rxvlan for them
+	 */
+	if (is_afvf(pcifunc))
+		return 0;
+
 	pfvf = rvu_get_pfvf(rvu, pcifunc);
 	if (pfvf->rxvlan)
 		return 0;
@@ -2533,14 +2538,6 @@ int rvu_mbox_handler_NIX_RXVLAN_ALLOC(struct rvu *rvu, struct msg_req *req,
 	if (err)
 		return err;
 
-	/* enable new entry */
-	ena_req.hdr.pcifunc = pcifunc;
-	ena_req.entry = alloc_rsp.entry_list[0];
-
-	err = rvu_mbox_handler_NPC_MCAM_ENA_ENTRY(rvu, &ena_req, rsp);
-	if (err)
-		goto free_entry;
-
 	/* update entry to enable rxvlan offload */
 	blkaddr = rvu_get_blkaddr(rvu, BLKTYPE_NIX, pcifunc);
 	if (blkaddr < 0) {
diff --git a/drivers/soc/marvell/octeontx2/rvu_npc.c b/drivers/soc/marvell/octeontx2/rvu_npc.c
index 800a0a6c61fc..e4ae7c2ed7a4 100644
--- a/drivers/soc/marvell/octeontx2/rvu_npc.c
+++ b/drivers/soc/marvell/octeontx2/rvu_npc.c
@@ -372,7 +372,7 @@ void rvu_npc_install_promisc_entry(struct rvu *rvu, u16 pcifunc,
 {
 	struct npc_mcam *mcam = &rvu->hw->mcam;
 	struct mcam_entry entry = { {0} };
-	struct nix_rx_action action;
+	struct nix_rx_action action = { };
 	int blkaddr, ucast_idx, index, kwi;
 
 	/* Only PF or AF VF can add a promiscuous entry */
-- 
2.17.1

